contrib/cgi/src/cgi.c

```diff
@@ -0,0 +1,683 @@
+#include <config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <gio/gio.h>
+#include <json-glib/json-glib.h>
+#include <rauc-installer-generated.h>
+
+/*
+ * Bundle/status file locations
+ * It is assumed that these files do not survive a reboot.
+ */
+const gchar *BUNDLE_TARGET_LOCATION = "/tmp/bundle.raucb";
+const gchar *STATUS_FILE_LOCATION = "/tmp/rauc-status.json";
+
+/*
+ * HTTP request headers that should be saved during PUT and are then available
+ * in status method, lowercase, NULL-terminated
+ */
+const gchar *HTTP_HEADERS[] = {"upload_client_id", NULL};
+
+
+#define CGI_ERROR cgi_error_quark()
+
+static GQuark cgi_error_quark(void)
+{
+	return g_quark_from_static_string("cgi_error_quark");
+}
+
+typedef enum {
+	CGI_ERROR_FAILED,
+	CGI_ERROR_METHOD_NOT_ALLOWED,
+	CGI_ERROR_LENGTH_REQUIRED,
+	CGI_ERROR_BAD_REQUEST,
+	CGI_ERROR_SERVICE_UNAVAILABLE,
+} CGIError;
+
+/*
+ * Print HTTP headers and two new lines to start HTTP body.
+ */
+static void print_headers(const gchar* status, const gchar* content_type)
+{
+	g_print("Status: %s\n", status);
+	g_print("Content-type: %s\n", content_type);
+	g_print("\n");
+}
+
+/*
+ * Retrieve LastError property via rauc D-Bus service.
+ */
+static gchar* get_last_error(RInstaller *installer, GError **error)
+{
+	gchar* last_error = r_installer_dup_last_error(installer);
+	if (!last_error)
+		g_set_error(error, CGI_ERROR, CGI_ERROR_FAILED, "D-Bus error");
+
+	return last_error;
+}
+
+/*
+ * Retrieve Operation property via rauc D-Bus service.
+ */
+static gchar* get_operation(RInstaller *installer, GError **error)
+{
+	gchar* operation = r_installer_dup_operation(installer);
+	if (!operation)
+		g_set_error(error, CGI_ERROR, CGI_ERROR_FAILED, "D-Bus error");
+
+	return operation;
+}
+
+/*
+ * Retrieve ProgressUpdated property via rauc D-Bus service.
+ */
+static gboolean get_progress_updated(RInstaller *installer, gint *percentage, gchar **description, GError **error)
+{
+	gboolean res = FALSE;
+	GVariant* progress_updated;
+	GVariant* g_percentage;
+	GVariant* g_description;
+
```

Comment:
Would be worth adding

```
        g_return_val_if_fail(percentage != NULL, FALSE);
        g_return_val_if_fail(description == NULL || *description == NULL, FALSE);
```

This also reveals that the `get_progress_updated()` method breaks the convention that GError functions returning false should have GError set.

Glib manual say:

> Avoid creating functions which have a boolean return value and a GError parameter, but where the boolean does something other than signal whether the GError is set. Among other problems, it requires C callers to allocate a temporary error. Instead, provide a "gboolean *" out parameter. There are functions in GLib itself such as g_key_file_has_key() that are deprecated because of this.

Reply:
> Avoid creating functions which have a boolean return value and a GError parameter, but where the boolean does something other than signal whether the GError is set.

If I am not mistaken that is actually what the function does. The function has a boolean return value and a GError parameter and the booleans does signal whether the GError is set. So what's wrong with that?

Reply:
Mh, I must have looked at another function while writing this...

Ah... it was the user of `get_progress_updated()` (I stumbled accross this when testing with `g_return_val_if_fail()`).

It is actually `get_will_reboot()` I was talking about.

Reply:
Adding..
```
g_return_val_if_fail(percentage != NULL, FALSE);
g_return_val_if_fail(description == NULL || *description == NULL, FALSE);
```
..here would break the mentioned rule from GLib manual pages..

> Avoid creating functions which have a boolean return value and a GError parameter, but where the boolean does something other than signal whether the GError is set.

Reply:
Yesss, but this is mainly for catching invalid API usage and not detecting and handling runtime errors. It does not signal a GError is set but at least an error condition. Let me say, if that produces an error, the missing content of GError is your smallest problem :laughing: 

As glib does it this way too I would expect it to be meant to do it this way.

Reply:
Alright.